### PR TITLE
Show blast error when individual jobs in a submission fails

### PR DESCRIPTION
**Pull request by jyothishnt**
## Related JIRA Issue(s)
https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-1772

## Deployment URL(s)
http://blast-individual-job-error.review.ensembl.org

## Views affected
Blast

1. If you force a job fail [here ](https://github.com/Ensembl/ensembl-client/pull/899/files#diff-4dcb50ac8116a9f28320c4d2f53071426b327b05801d55997451161e446df1e8R124) after fetching the results, you should be able to see the failed job status with tooltip.
2. Please use this branch on the sandbox for testing

### CODE REVIEW

File: src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx [hunk 1]
```diff
@@ -39,6 +40,10 @@ export type Props = {
   submission: BlastSubmission;
 };
 
+export const FailedSubmissionHelpText = () => (
+  <p>Unable to get results for this sequence</p>
+);
```
Comment (azangru) [inline src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx:45]: When a submission is collapsed, the text doesn't really make sense:

image

We may choose to:
- keep this text in the collapsed view regardless (and wait for Andrea to come back and correct it)
- replace it with a more appropriate text that we write ourselves
- remove the question button in the collapsed view (until Andrea comes back and corrects it)

File: src/content/app/tools/blast/views/blast-submission-results/components/blast-results-per-sequence/BlastResultsPerSequence.tsx [hunk 2]
```diff
@@ -64,72 +66,97 @@ const BlastResultsPerSequence = (props: BlastResultsPerSequenceProps) => {
...
-        <div>Sequence {sequence.id}</div>
-        <div className={styles.sequenceHeader}>
-          <div className={headerClasses}>
-            <ShowHide
-              label={sequenceHeaderLabel}
-              isExpanded={shouldShowParamaters}
-              onClick={() => showParamaters(!shouldShowParamaters)}
-            ></ShowHide>
-            {shouldUseGenomicHitsDiagram && <BlastGenomicHitsDiagramLegend />}
+  const hasAllJobsFailed = blastResults.every(
```
Comment (azangru) [inline src/content/app/tools/blast/views/blast-submission-results/components/blast-results-per-sequence/BlastResultsPerSequence.tsx:69]: `has` -> `have`?

File: src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx [hunk 3]
```diff
@@ -162,11 +167,12 @@ const StatusElement = ({ jobs }: { jobs: BlastJob[] }) => {
...
-    if (jobs.length === 1 || jobs.every((job) => job.status === 'FAILURE')) {
-      return <span className={elementClasses}>Failed</span>;
-    } else {
-      return <span className={elementClasses}>Some jobs failed</span>;
-    }
+    return (
+      <div>
+        <span className={elementClasses}> Job failed </span>
+        <QuestionButton helpText={<FailedSubmissionHelpText />} />
+      </div>
```
Comment (azangru) [inline src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx:174]: You probably want to adjust the vertical alignment here:

image

File: src/content/app/tools/blast/views/blast-submission-results/BlastSubmissionResults.tsx [hunk 4]
```diff
@@ -121,6 +121,10 @@ const Main = () => {
     })
     .filter((job) => !!job.data) as BlastJobWithResults[]; // only care about BLAST jobs that have the results from the server; they all should do if the apis behave properly
 
+  // Add below line to force a failure
+  // allJobResultsWithData[0].status="FAILURE";
+  // allJobResultsWithData[1].status="FAILURE";
```
Comment (azangru) [inline src/content/app/tools/blast/views/blast-submission-results/BlastSubmissionResults.tsx:126]: Remember to delete before merging...

File: src/content/app/tools/blast/views/blast-submission-results/components/blast-results-per-sequence/BlastResultsPerSequence.tsx [hunk 5]
```diff
@@ -64,72 +66,97 @@ const BlastResultsPerSequence = (props: BlastResultsPerSequenceProps) => {
...
-          {shouldShowParamaters && (
-            <JobParameters
-              sequenceValue={sequence.value}
-              parameters={parameters}
-              preset={preset}
-            />
-          )}
         </div>
+      </div>
+    );
```
[the remark this answers is not in the document]
Comment (azangru) [reply src/content/app/tools/blast/views/blast-submission-results/components/blast-results-per-sequence/BlastResultsPerSequence.tsx:91]: You can see that another XD [partly contradicts](https://xd.adobe.com/view/47c05579-d7aa-414a-a0ca-e2b5e2c88636-1293/screen/88c148e3-c947-4203-b1ca-fcd52901c4a4?fullscreen) this layout — in that you can access the submission parameters there.

File: src/content/app/tools/blast/views/blast-submission-results/components/blast-results-per-sequence/BlastResultsPerSequence.tsx [hunk 6]
```diff
@@ -64,72 +66,97 @@ const BlastResultsPerSequence = (props: BlastResultsPerSequenceProps) => {
...
-              label={sequenceHeaderLabel}
-              isExpanded={shouldShowParamaters}
-              onClick={() => showParamaters(!shouldShowParamaters)}
-            ></ShowHide>
-            {shouldUseGenomicHitsDiagram && <BlastGenomicHitsDiagramLegend />}
+  const hasAllJobsFailed = blastResults.every(
+    (job) => job.status === 'FAILURE'
+  );
+
+  if (hasAllJobsFailed) {
```
Comment (azangru) [inline src/content/app/tools/blast/views/blast-submission-results/components/blast-results-per-sequence/BlastResultsPerSequence.tsx:73]: - What happens if there is only one job per sequence?
- I think you are only covering [this XD](https://xd.adobe.com/view/47c05579-d7aa-414a-a0ca-e2b5e2c88636-1293/screen/566d3729-59d6-444a-8774-66d84af759dc?fullscreen), where all jobs per sequence have failed. But there is [another XD](https://xd.adobe.com/view/47c05579-d7aa-414a-a0ca-e2b5e2c88636-1293/screen/88c148e3-c947-4203-b1ca-fcd52901c4a4?fullscreen), where only some jobs per sequence have failed

File: src/content/app/tools/blast/views/blast-submission-results/components/blast-results-per-sequence/BlastResultsPerSequence.tsx [hunk 7]
```diff
@@ -64,72 +66,97 @@ const BlastResultsPerSequence = (props: BlastResultsPerSequenceProps) => {
...
+          <div>Sequence {sequence.id}</div>
+          <div className={styles.sequenceHeader}>
+            <div className={headerClasses}>{sequenceHeaderLabel}</div>
+          </div>
+          <div>
+            <span className={styles.againstText}>Against</span>{' '}
+            <span>{species.length} species</span>
+          </div>
+          <div>
+            <span className={styles.failedJobStatus}> Job failed </span>
```
Comment (imransl) [inline src/content/app/tools/blast/views/blast-submission-results/components/blast-results-per-sequence/BlastResultsPerSequence.tsx:86]: Why are there whitespaces on each side of the text here?

File: src/content/app/tools/blast/views/blast-submission-results/components/blast-results-per-sequence/BlastResultsPerSequence.tsx [hunk 8]
```diff
@@ -64,72 +66,97 @@ const BlastResultsPerSequence = (props: BlastResultsPerSequenceProps) => {
...
+      <div className={styles.wrapper}>
+        <div className={styles.resultsSummaryRow}>
+          <div>Sequence {sequence.id}</div>
+          <div className={styles.sequenceHeader}>
+            <div className={headerClasses}>
+              <ShowHide
+                label={sequenceHeaderLabel}
+                isExpanded={shouldShowParamaters}
+                onClick={() => showParamaters(!shouldShowParamaters)}
+              ></ShowHide>
```
Comment (imransl) [inline src/content/app/tools/blast/views/blast-submission-results/components/blast-results-per-sequence/BlastResultsPerSequence.tsx:103]: I am surprised Prettier didn't change this into a self closing tag

File: src/content/app/tools/blast/views/blast-submission-results/components/blast-results-per-sequence/BlastResultsPerSequence.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (ens-st3) [reply src/content/app/tools/blast/views/blast-submission-results/components/blast-results-per-sequence/BlastResultsPerSequence.tsx:91]: Does feel that it could be improved by being able to view the details

File: src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx (hunk 1 above)
Comment (ens-st3) [reply src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx:45]: Question - under what circumstances can, in one submission, one or more sequences fail and one or more succeed. I can see how we might get no alignments for the one or more sequences, but I don't know what could cause one or more to fail

Comment (azangru) [reply src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx:45]: I don't have a rational answer. Those hypothetical errors are probably the most improbable, especially if it's for several sequence-species combinations, like in [this XD](https://xd.adobe.com/view/47c05579-d7aa-414a-a0ca-e2b5e2c88636-1293/screen/566d3729-59d6-444a-8774-66d84af759dc?fullscreen) (1 sequence against 2 species). But then, shouldn't we proceed from the assumption that any job, for whatever reason, however rarely, might fail?

Comment (ens-st3) [reply src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx:45]: We are discussing how to deal with something that we have no reason to expect to happen. Could / should we have a generic way to deal with these hypothetical situations (I agree we need something) but focus more on the situations that we know will fail ?

Comment (azangru) [reply src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx:45]: These are questions for Andrea. So far we've got designs for the following scenarios:

1) Something happened to a submission as a whole ([link](https://xd.adobe.com/view/47c05579-d7aa-414a-a0ca-e2b5e2c88636-1293/screen/0070e1b1-554a-4f96-afd0-4d620e592019?fullscreen))

image

2) Something happened to one or more jobs grouped by a sequence, and represented visually as a box ([link](https://xd.adobe.com/view/47c05579-d7aa-414a-a0ca-e2b5e2c88636-1293/screen/ca48bfae-0a80-4d9d-b8d9-133f3c8db809?fullscreen), although see [another link](https://xd.adobe.com/view/47c05579-d7aa-414a-a0ca-e2b5e2c88636-1293/screen/3d44f722-8823-417f-ba94-ff02b9144f0e?fullscreen) as a variation):

image

3) Something happened to one or more jobs inside of an expanded box ([link](https://xd.adobe.com/view/47c05579-d7aa-414a-a0ca-e2b5e2c88636-1293/screen/88c148e3-c947-4203-b1ca-fcd52901c4a4?fullscreen)):

image

Looks pretty generic to me.

File: src/content/app/tools/blast/components/blast-settings/BlastSettings.scss [hunk 9]
```diff
@@ -35,17 +35,17 @@
...
 .blastSubmissionName label span,
 .select label span {
   font-weight: $light;
   line-height: 27px;
   justify-self: end;
 }
 
+.parametersColumn .matrixSetting .select label {
+  grid-template-columns: 140px 110px;
+}
```
Comment (azangru) [inline src/content/app/tools/blast/components/blast-settings/BlastSettings.scss:47]: Nothing seems to have changed about this CSS block? Did you move it because stylelint was complaining about styles order, or because you wanted to change its CSS rules, but then decided not to? If it's the latter, could you please revert this change?

File: src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx [hunk 10]
```diff
@@ -143,30 +146,52 @@ const SequenceBox = (props: SequenceBoxProps) => {
...
 
-const StatusElement = ({ jobs }: { jobs: BlastJob[] }) => {
-  const hasRunningJobs = jobs.some((job) => job.status === 'RUNNING');
-  const hasFailedJobs = jobs.some((job) => job.status === 'FAILURE');
+type StatusElementProps = {
+  sequenceId?: number;
+  jobs: BlastJob[];
+};
+
+export const StatusElement = (props: StatusElementProps) => {
```
Comment (azangru) [inline src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx:164]: `StatusElement` -> `JobsStatus`? The "element" part in a component name doesn't tell us anything.

File: src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx [hunk 11]
```diff
@@ -143,30 +146,52 @@ const SequenceBox = (props: SequenceBoxProps) => {
...
+  const { sequenceId, jobs } = props;
+
+  isFailedBlastSubmission();
+
+  const hasRunningJobs = jobs.some(
+    (job) =>
+      job.status === 'RUNNING' &&
+      (sequenceId ? sequenceId === job.sequenceId : true)
+  );
+  const hasAllFailedJobs = jobs.every(
```
Comment (azangru) [inline src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx:174]: maybe `hasAllFailedJobs` -> `hasOnlyFailedJobs`? Although I am not feeling strongly about this one.

File: src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx [hunk 12]
```diff
@@ -143,30 +146,52 @@ const SequenceBox = (props: SequenceBoxProps) => {
...
+
+export const StatusElement = (props: StatusElementProps) => {
+  const { sequenceId, jobs } = props;
+
+  isFailedBlastSubmission();
+
+  const hasRunningJobs = jobs.some(
+    (job) =>
+      job.status === 'RUNNING' &&
+      (sequenceId ? sequenceId === job.sequenceId : true)
```
Comment (azangru) [inline src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx:172]: Why are we doing this sequence id check in this component? Why are we not passing it only relevant jobs from the parent?

File: src/content/app/tools/blast/views/blast-submission-results/components/blast-results-per-sequence/BlastResultsPerSequence.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (azangru) [reply src/content/app/tools/blast/views/blast-submission-results/components/blast-results-per-sequence/BlastResultsPerSequence.tsx:87]: Reviewing on Jan 12. Didn't we say that we are allowing the user to view submission parameters? Or did we change our mind?

image

File: src/content/app/tools/blast/views/blast-submission-results/BlastSubmissionResults.scss [hunk 13]
```diff
@@ -1,6 +1,6 @@
 @import 'src/styles/common';
 
-$container_max_width: 1860px;
+$container_max_width: 1770px;
```
Comment (azangru) [inline src/content/app/tools/blast/views/blast-submission-results/BlastSubmissionResults.scss:3]: Not sure if for this PR or for another one; but could we do it as a `calc` of maximum BLAST content width (or whatever should be the name of the 1800px that we use for the BLAST navigation in the top bar) minus the standard gutter that we have on the left? Otherwise, 1770px is a bit magical.

File: src/content/app/tools/blast/views/blast-submission-results/components/blast-results-per-sequence/BlastResultsPerSequence.scss [hunk 14]
```diff
@@ -59,3 +59,13 @@
...
+
+.collapsedJobStatus {
+  justify-self: end;
+  color: $red;
+  padding-right: 20px;
+}
+
+.failedJobStatus {
+  color: $red;
+}
```
Comment (azangru) [inline src/content/app/tools/blast/views/blast-submission-results/components/blast-results-per-sequence/BlastResultsPerSequence.scss:71]: Is the job status always red? Would it make sense to have a `.jobStatus` (or `.faliedJob`) class that says that its color is red, and then either a `.collapsedJobStatus` class or (even better) a `.resultsSummaryRow .jobStatus` combo that sets `justify-self` and `padding-right`?

File: src/content/app/tools/blast/views/blast-submission-results/components/blast-results-per-sequence/BlastResultsPerSequence.tsx [hunk 15]
```diff
@@ -23,6 +23,7 @@ import FeatureLengthRuler from 'src/shared/components/feature-length-ruler/Featu
 import JobParameters from '../job-parameters/JobParameters';
 import SingleBlastJobResult from '../single-blast-job-result/SingleBlastJobResult';
 import { BlastGenomicHitsDiagramLegend } from 'src/content/app/tools/blast/components/blast-genomic-hits-diagram';
+import { StatusElement } from '../../../../components/listed-blast-submission/ListedBlastSubmission';
```
Comment (azangru) [inline src/content/app/tools/blast/views/blast-submission-results/components/blast-results-per-sequence/BlastResultsPerSequence.tsx:26]: `../../../../` = 😞

File: src/content/app/tools/blast/views/blast-submission-results/components/single-blast-job-result/SingleBlastJobResult.tsx [hunk 16]
```diff
@@ -182,38 +182,52 @@ const SingleBlastJobResult = (props: SingleBlastJobResultProps) => {
...
+  if (jobResult.status === 'FAILURE') {
+    return (
+      <div className={styles.failedJobSummaryRow}>
+        <div className={styles.failedJobStatus}>Job failed</div>
+        <BlastSpecies
+          species={speciesInfo}
+          isExpanded={false}
+          toggleExpanded={setExpanded}
+          jobResult={jobResult}
+        />
```
Comment (azangru) [inline src/content/app/tools/blast/views/blast-submission-results/components/single-blast-job-result/SingleBlastJobResult.tsx:194]: image

File: src/content/app/tools/blast/views/blast-submission-results/components/single-blast-job-result/SingleBlastJobResult.scss [hunk 17]
```diff
@@ -28,6 +28,30 @@
...
+  }
+
+  &:last-of-type {
+    padding-bottom: 46px;
+  }
+
+  .blastSpecies {
+    color: $dark-grey;
+  }
+}
```
Comment (azangru) [inline src/content/app/tools/blast/views/blast-submission-results/components/single-blast-job-result/SingleBlastJobResult.scss:53]: Wondering if we can just add an extra class to the corresponding element (`.resultsSummaryRow.failedJobRow`). so that all the padding and stuff comes from `.resultsSummaryRow` (rather than being copy-pasted), and only the grid settings are updated by a `failedJobRow` class

File: src/content/app/tools/blast/components/blast-settings/BlastSettings.scss (hunk 9 above)
Comment (jyothishnt) [reply src/content/app/tools/blast/components/blast-settings/BlastSettings.scss:47]: Yes because of the linter complaining about its order.

File: src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx [hunk 18]
```diff
@@ -143,30 +146,52 @@ const SequenceBox = (props: SequenceBoxProps) => {
...
+  );
+  const hasAllFailedJobs = jobs.every(
+    (job) =>
+      job.status === 'FAILURE' &&
+      (sequenceId ? sequenceId === job.sequenceId : true)
+  );
+  const hasPartFailedJobs = jobs.some(
```
[the remark this answers is not in the document]
Comment (jyothishnt) [reply src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx:179]: I think it makes more sense because we use the same terminology in the UI and code, makes life easier and have same meaning too.

File: src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx (hunk 11 above)
Comment (jyothishnt) [reply src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx:174]: or `hasAllJobsFailed` ? :-)

File: src/content/app/tools/blast/views/blast-submission-results/BlastSubmissionResults.scss (hunk 13 above)
Comment (jyothishnt) [reply src/content/app/tools/blast/views/blast-submission-results/BlastSubmissionResults.scss:3]: Its a minor change, doesnt require another PR.

File: src/content/app/tools/blast/views/blast-submission-results/components/blast-results-per-sequence/BlastResultsPerSequence.scss (hunk 14 above)
Comment (jyothishnt) [reply src/content/app/tools/blast/views/blast-submission-results/components/blast-results-per-sequence/BlastResultsPerSequence.scss:71]: Confused, lets discuss this over code review

File: src/content/app/tools/blast/views/blast-submission-results/components/single-blast-job-result/SingleBlastJobResult.tsx (hunk 16 above)
Comment (jyothishnt) [reply src/content/app/tools/blast/views/blast-submission-results/components/single-blast-job-result/SingleBlastJobResult.tsx:194]: This is not to worry. Its because your results data:[] is not empty. The code current code handles it and wont be displayed when the data is empty.

File: src/content/app/tools/blast/views/blast-submission-results/components/single-blast-job-result/SingleBlastJobResult.scss (hunk 17 above)
Comment (jyothishnt) [reply src/content/app/tools/blast/views/blast-submission-results/components/single-blast-job-result/SingleBlastJobResult.scss:53]: lets discuss this together with the above css suggestion.

File: src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx (hunk 18 above)
[the remark this answers is not in the document]
Comment (azangru) [reply src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx:177]: One job in our code is one sequence against one species. It can never part-fail. UI is for Andrea.

File: src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx (hunk 11 above)
Comment (azangru) [reply src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx:172]: > or hasAllJobsFailed ?

I don't know what that means — does it mean that (1) something (e.g. a box) only **has** jobs that failed, or (2) that all jobs **have** failed? From the word order, it sounds like the second; but then the verb needs to be different.

File: src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx (hunk 12 above)
Comment (jyothishnt) [reply src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx:170]: Removed them as we dont need to do this check.

File: src/content/app/tools/blast/BlastPage.scss [hunk 19]
```diff
@@ -1,6 +1,7 @@
 @import 'src/styles/common';
 
 .blastPage {
+  --blast-container-max-width: 1800px;
```
Comment (azangru) [inline src/content/app/tools/blast/BlastPage.scss:4]: There are a couple of other places where the magic 1800px value is used. Some of them, on our BLAST pages:

image

While we can't use CSS variables in media queries in `_breakpoints.scss` and `globalConfig.ts`, we should probably update other BLAST SCSS files to use the CSS variable above.

File: src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx [hunk 20]
```diff
@@ -143,30 +146,37 @@ const SequenceBox = (props: SequenceBoxProps) => {
...
 
-const StatusElement = ({ jobs }: { jobs: BlastJob[] }) => {
+type JobStatusProps = {
+  jobs: BlastJob[];
+};
+
+export const JobStatus = (props: JobStatusProps) => {
+  const { jobs } = props;
+
+  isFailedBlastSubmission();
```
Comment (azangru) [inline src/content/app/tools/blast/components/listed-blast-submission/ListedBlastSubmission.tsx:164]: What's happening here?

**Review by azangru — APPROVED** [review]
